lib/pbench/agent/tool_data_sink.py

```diff
@@ -2060,7 +2062,9 @@ def start(prog: Path, parsed: Arguments):
         # E.g. params = '{ "channel_prefix": "some-prefix",
         #                  "benchmark_run_dir": "/loo/goo" }'
         params = json.loads(params_str)
-        ToolDataSink.fetch_params(params, pbench_run)
+
+        benchmark_run_dir = BenchmarkRunDir(params["benchmark_run_dir"], pbench_run)
```

Comment:
I don't want to gloss over one "rough edge" here relative to the way Peter had structured them: if `"benchmark_run_dir"` is not present in the JSON object, then we'll get a `KeyError` raised here rather than a `ToolDataSinkError`.

The error will be handled with reasonable grace either way, and I think the slight inconsistency is worth the increased simplicity in the code. @portante, if you disagree, I can add either a try block or an explicit key-check before accessing params. However, I think we can just ensure that the key will be there via testing.

Reply:
Why not just have the `ToolDataSinkParams` field be a string, and then let the `ToolDataSink` constructor convert it to a `BenchmarkRunDir` object?

Reply:
@portante, I'm not sure what you're asking.

I think that "`ToolDataSinkParams` as a string" is the value in `params` here; passing that around creates the problem that I was trying to solve in this PR:  we want to validate it here, and the best way to validate it is to construct the `ToolDataSinkParams` object; if we defer that construction, then we have to come up with a separate way to validate it here (and constructing it twice seems silly for a couple of reasons).

Also, the instantiation of the `BenchmarkRunDir` object requires the value of `pbench_run`; if we do that here, then we don't have to pass that value around, either (and we find out up front if there's a problem with it).

So, I think this is the best location for instantiating both the `BenchmarkRunDir` and the `ToolDataSinkParams` objects.  The lingering question is, if the instantiation of the `BenchmarkRunDir` fails, how do we want to see that reported?...does it need to be a `ToolDataSinkError` or is some other error (such as `KeyError`) acceptable?

Reply:
Let me get this PR split, first.  Then I can look at that.  🥴